File: python/xpict/edge_plan.py
"""EdgePlan build + host ``coord_gen`` processor (Python / RDKit)."""

from __future__ import annotations

from typing import Any

from xpict.contracts.edge import (
    AlignOpts,
    CoordGenMoleculeResult,
    CoordGenTask,
    CoordGenTaskResult,
    EdgePlan,
    EdgeResult,
    MolTemplate,
)

# Keep in sync with ``xpict_core::edge::MIN_MCS_ATOMS``.
_MIN_MCS_ATOMS = 3


def validate_edge_plan(plan: EdgePlan | dict[str, Any]) -> EdgePlan:
    """Structural checks via Rust core (unique ids, roots, structure fields)."""
    from xpict.native_bridge import validate_edge_plan as _core_validate

    return EdgePlan.model_validate(_core_validate(plan))


def build_align_plan(
    *,
    template_source: str,
    query_source: str,
    atom_map: list[tuple[int, int]] | None = None,
    template_id: str = "m_0",
    query_id: str = "m_1",
) -> EdgePlan:
    """Two-node forest: template root + one query child (simple-client helper)."""
    child_align = AlignOpts(atom_map=atom_map) if atom_map is not None else AlignOpts()
    return EdgePlan(
        tasks=[
            CoordGenTask(
                roots=[
                    MolTemplate(
                        id=template_id,
                        smiles=template_source,
                        align=None,
                        template_for=[
                            MolTemplate(
                                id=query_id,
                                smiles=query_source,
                                align=child_align,
                                template_for=[],
                            )
                        ],
                    )
                ]
            )
        ]
    )


def _source_of(node: MolTemplate) -> str:
    for v in (node.smiles, node.cxsmiles, node.molfile):
        if v is not None and str(v).strip():
            return str(v)
    raise ValueError(f"MolTemplate {node.id}: empty structure")


def _layout(
    source: str,
    *,
    template: str | None,
    atom_map: list[tuple[int, int]] | None,
    id: str | None,
    min_atoms: int,
) -> tuple[dict[str, Any], str, str, list[tuple[int, int]] | None]:
    """Returns molecule, pose_molblock, method, used_map."""
    from xpict.client import layout_with_rdkit

    molecule, pose, meta = layout_with_rdkit(
        source,
        template=template,
        id=id,
        atom_map=atom_map,
        min_atoms=min_atoms,
    )
    return molecule, pose, meta["method"], meta.get("used_map")


def process_edge_plan(plan: EdgePlan | dict[str, Any]) -> EdgeResult:
    """Run all ``coord_gen`` tasks; always return a flat molecule list.

    Align failure → automatic free (unaligned) layout for that mol; ``method``
    is ``none`` and ``ok`` stays true when coords were produced.
    """
    p = validate_edge_plan(plan)
    task_results: list[CoordGenTaskResult] = []

    for task in p.tasks:
        rows: list[CoordGenMoleculeResult] = []
        poses: dict[str, str] = {}

        def visit(node: MolTemplate, parent_id: str | None) -> None:
            source = _source_of(node)
            min_atoms = _MIN_MCS_ATOMS
            atom_map = None
            if node.align is not None:
                if node.align.min_atoms is not None:
                    min_atoms = int(node.align.min_atoms)
                atom_map = node.align.atom_map

            template_pose = poses.get(parent_id) if parent_id else None

            def free_layout() -> tuple[dict[str, Any], str]:
                molecule, pose, _method, _used = _layout(
                    source,
                    template=None,
                    atom_map=None,
                    id=node.id,
                    min_atoms=min_atoms,
                )
                return molecule, pose

            try:
                if parent_id is None:
                    molecule, pose = free_layout()
                    rows.append(
                        CoordGenMoleculeResult(
                            id=node.id,
                            ok=True,
                            method="free",
                            used_map=None,
                            molecule=molecule,
                            error=None,
                        )
                    )
                    poses[node.id] = pose
                else:
                    if template_pose is None:
                        raise RuntimeError(f"missing parent pose {parent_id}")
                    molecule, pose, method, used = _layout(
                        source,
                        template=template_pose,
                        atom_map=atom_map,
                        id=node.id,
                        min_atoms=min_atoms,
                    )
                    if method in ("atom_map", "mcs"):
                        rows.append(
                            CoordGenMoleculeResult(
                                id=node.id,
                                ok=True,
                                method=method,  # type: ignore[arg-type]
                                used_map=used,
                                molecule=molecule,
                                error=None,
                            )
                        )
                    else:
                        # Align soft-fail already returned free coords from layout;
                        # re-run unaligned to guarantee a clean free pose.
                        molecule, pose = free_layout()
                        rows.append(
                            CoordGenMoleculeResult(
                                id=node.id,
                                ok=True,
                                method="none",
                                used_map=None,
                                molecule=molecule,
                                error="align failed; fell back to unaligned coord gen",
                            )
                        )
                    poses[node.id] = pose
            except Exception as e:  # noqa: BLE001 — try free fallback, then record
                try:
                    molecule, pose = free_layout()
                    rows.append(
                        CoordGenMoleculeResult(
                            id=node.id,
                            ok=True,
                            method="none" if parent_id is not None else "free",
                            used_map=None,
                            molecule=molecule,
                            error=(
                                f"align/layout error ({e}); fell back to unaligned coord gen"
                                if parent_id is not None
                                else None
                            ),
                        )
                    )
                    poses[node.id] = pose
                except Exception as e2:  # noqa: BLE001
                    rows.append(
                        CoordGenMoleculeResult(
                            id=node.id,
                            ok=False,
                            method="none",
                            used_map=None,
                            molecule=None,
                            error=str(e2),
                        )
                    )
                    return

            for child in node.template_for:
                visit(child, node.id)

        for root in task.roots:
            visit(root, None)

        task_ok = all(r.ok for r in rows)
        task_results.append(
            CoordGenTaskResult(ok=task_ok, molecules=rows)
        )

    return EdgeResult(results=task_results)

File: crates/xpict/src/edge_plan.rs
//! Process [`xpict_core::EdgePlan`] with native RDKit Depictor.

use std::collections::HashMap;

use xpict_core::edge::{
    AlignOpts, CoordGenMoleculeResult, CoordMethod, EdgePlan, EdgeResult, EdgeTask,
    EdgeTaskResult, MolTemplate, MIN_MCS_ATOMS,
};

use crate::layout::layout_with_rdkit_meta;
use crate::Error;

/// Build a two-node align forest (template root + query child).
pub fn build_align_plan(
    template_source: &str,
    query_source: &str,
    atom_map: Option<Vec<(u32, u32)>>,
) -> EdgePlan {
    EdgePlan::new_v1(vec![EdgeTask::CoordGen {
        roots: vec![MolTemplate {
            id: "m_0".into(),
            smiles: Some(template_source.into()),
            cxsmiles: None,
            molfile: None,
            align: None,
            template_for: vec![MolTemplate {
                id: "m_1".into(),
                smiles: Some(query_source.into()),
                cxsmiles: None,
                molfile: None,
                align: Some(AlignOpts {
                    atom_map,
                    min_atoms: None,
                }),
                template_for: vec![],
            }],
        }],
    }])
}

fn source_of(node: &MolTemplate) -> Result<String, Error> {
    node.source()
        .map(str::to_string)
        .map_err(Error::Parse)
}

/// Run all ``coord_gen`` tasks; always return flat molecule rows.
///
/// Align failure → automatic free (unaligned) layout; ``method: None``,
/// ``ok: true`` when coords were produced.
pub fn process_edge_plan(plan: &EdgePlan) -> Result<EdgeResult, Error> {
    Ok(process_edge_plan_with_frames(plan)?.0)
}

/// Like [`process_edge_plan`], plus pose molblocks keyed by mol id (for
/// [`crate::Rendered::frame_molblock`]).
pub fn process_edge_plan_with_frames(
    plan: &EdgePlan,
) -> Result<(EdgeResult, HashMap<String, String>), Error> {
    plan.validate().map_err(Error::Parse)?;
    let mut task_results = Vec::new();
    let mut all_poses: HashMap<String, String> = HashMap::new();

    for task in &plan.tasks {
        match task {
            EdgeTask::CoordGen { roots } => {
                let mut rows: Vec<CoordGenMoleculeResult> = Vec::new();
                let mut poses: HashMap<String, String> = HashMap::new();

                for root in roots {
                    visit(root, None, &mut rows, &mut poses)?;
                }

                all_poses.extend(poses);
                let ok = rows.iter().all(|r| r.ok);
                task_results.push(EdgeTaskResult::CoordGen {
                    ok,
                    molecules: rows,
                });
            }
        }
    }

    Ok((EdgeResult::new_v1(task_results), all_poses))
}

fn free_layout(
    source: &str,
    id: &str,
) -> Result<(xpict_core::MoleculeIn, String), Error> {
    let (molecule, pose, _) =
        layout_with_rdkit_meta(source, None, None, Some(id.to_string()))?;
    Ok((molecule, pose))
}

fn visit(
    node: &MolTemplate,
    parent_id: Option<&str>,
    rows: &mut Vec<CoordGenMoleculeResult>,
    poses: &mut HashMap<String, String>,
) -> Result<(), Error> {
    let source = source_of(node)?;
    let min_atoms = node
        .align
        .as_ref()
        .and_then(|a| a.min_atoms)
        .unwrap_or(MIN_MCS_ATOMS);
    let atom_map = node.align.as_ref().and_then(|a| a.atom_map.clone());

    match parent_id {
        None => match free_layout(&source, &node.id) {
            Ok((molecule, pose)) => {
                rows.push(CoordGenMoleculeResult {
                    id: node.id.clone(),
                    ok: true,
                    method: CoordMethod::Free,
                    used_map: None,
                    molecule: Some(molecule),
                    error: None,
                });
                poses.insert(node.id.clone(), pose);
            }
            Err(e) => {
                rows.push(CoordGenMoleculeResult {
                    id: node.id.clone(),
                    ok: false,
                    method: CoordMethod::None,
                    used_map: None,
                    molecule: None,
                    error: Some(e.to_string()),
                });
                return Ok(());
            }
        },
        Some(pid) => {
            let template = match poses.get(pid) {
                Some(p) => p.clone(),
                None => {
                    // No parent pose — still try free layout.
                    match free_layout(&source, &node.id) {
                        Ok((molecule, pose)) => {
                            rows.push(CoordGenMoleculeResult {
                                id: node.id.clone(),
                                ok: true,
                                method: CoordMethod::None,
                                used_map: None,
                                molecule: Some(molecule),
                                error: Some(format!(
                                    "missing parent pose {pid}; fell back to unaligned coord gen"
                                )),
                            });
                            poses.insert(node.id.clone(), pose);
                        }
                        Err(e) => {
                            rows.push(CoordGenMoleculeResult {
                                id: node.id.clone(),
                                ok: false,
                                method: CoordMethod::None,
                                used_map: None,
                                molecule: None,
                                error: Some(e.to_string()),
                            });
                        }
                    }
                    for child in &node.template_for {
                        visit(child, Some(node.id.as_str()), rows, poses)?;
                    }
                    return Ok(());
                }
            };
            let map_ref = atom_map.as_deref();
            let map_ok_len = atom_map
                .as_ref()
                .map(|m| m.len() as u32 >= min_atoms)
                .unwrap_or(true);
            let use_map = if map_ok_len { map_ref } else { None };

            match layout_with_rdkit_meta(
                &source,
                Some(&template),
                use_map,
                Some(node.id.clone()),
            ) {
                Ok((molecule, pose, matched)) if matched => {
                    let method = if atom_map.is_some() {
                        CoordMethod::AtomMap
                    } else {
                        CoordMethod::Mcs
                    };
                    rows.push(CoordGenMoleculeResult {
                        id: node.id.clone(),
                        ok: true,
                        method,
                        used_map: atom_map.clone(),
                        molecule: Some(molecule),
                        error: None,
                    });
                    poses.insert(node.id.clone(), pose);
                }
                Ok(_) | Err(_) => {
                    // Align failed or errored → free layout fallback.
                    match free_layout(&source, &node.id) {
                        Ok((molecule, pose)) => {
                            rows.push(CoordGenMoleculeResult {
                                id: node.id.clone(),
                                ok: true,
                                method: CoordMethod::None,
                                used_map: None,
                                molecule: Some(molecule),
                                error: Some(
                                    "align failed; fell back to unaligned coord gen".into(),
                                ),
                            });
                            poses.insert(node.id.clone(), pose);
                        }
                        Err(e) => {
                            rows.push(CoordGenMoleculeResult {
                                id: node.id.clone(),
                                ok: false,
                                method: CoordMethod::None,
                                used_map: None,
                                molecule: None,
                                error: Some(e.to_string()),
                            });
                            return Ok(());
                        }
                    }
                }
            }
        }
    }

    for child in &node.template_for {
        visit(child, Some(node.id.as_str()), rows, poses)?;
    }
    Ok(())
}
